dvc/remote/index.py

```diff
@@ -0,0 +1,257 @@
+    )
+    for checksums in chunks(1024, data_checksums):
+        data = unhexlify("".join(checksums))
+        crc = zlib.crc32(data, crc)
+        compress_len += fobj.write(compress.compress(data))
+    compress_len += fobj.write(compress.flush())
+    endpos = fobj.tell()
+    fobj.seek(pos + _header_v1.size)
+    fobj.write(_compress_crc_v1.pack(compress_len, crc))
+    fobj.seek(endpos)
+
+
+def load(fobj, protocol=None, dir_suffix=".dir"):
+    """Read index checksums from the open file object ``fobj``.
+
+    Returns a 2-tuple of (dir_checksums, file_checksums).
+    """
+    data = fobj.read(4)
+    (protocol,) = struct.unpack("<I", data)
+    _verify_protocol(protocol)
+    fobj.seek(0)
+    return _load_v1(fobj, dir_suffix)
+
+
+def _load_v1(fobj, dir_suffix=""):
+    try:
+        protocol, dir_count, file_count = _header_v1.unpack(
+            fobj.read(_header_v1.size)
+        )
+        compress_len, file_crc = _compress_crc_v1.unpack(
+            fobj.read(_compress_crc_v1.size)
+        )
+    except struct.error as exc:
+        raise DvcException("Invalid v1 remote index file: {}".format(exc))
+    dir_checksums = set()
+    file_checksums = set()
+
+    def read_chunks():
+        # decompress 4kB chunks at a time and return uncompressed data aligned
+        # to a 4-byte (checksum length) boundary
+        decompress = zlib.decompressobj()
+        bytes_read = 0
+        unconsumed = b""
+        extra = b""
+        crc = 0
+        while bytes_read < compress_len:
+            compress_data = fobj.read(4096)
+            bytes_read += len(compress_data)
+            data = extra + decompress.decompress(unconsumed + compress_data)
+            unconsumed = decompress.unconsumed_tail
+            extra_count = len(data) % 4
+            extra = data[len(data) - extra_count :]
+            data = data[: len(data) - extra_count]
+            crc = zlib.crc32(data, crc)
+            yield data
+        data = extra + decompress.flush()
+        crc = zlib.crc32(data, crc)
+        if crc != file_crc:
+            raise DvcException("Remote index file failed CRC check")
+        return data
+
+    for data in read_chunks():
+        checksums = chunks(32, data.hex())
+        if len(dir_checksums) < dir_count:
+            dirs, files = split_at(dir_count - len(dir_checksums), checksums)
+            dir_checksums.update(checksum + dir_suffix for checksum in dirs)
+            file_checksums.update(files)
+        else:
+            file_checksums.update(checksums)
+
+    return dir_checksums, file_checksums
+
+
+class RemoteIndex(object):
+    """Class for locally indexing remote checksums.
+
+    Args:
+        repo: repo for this remote index.
+        name: name for this index. If name is provided, this index will be
+            loaded from and saved to ``.dvc/tmp/index/{name}.idx``.
+            If name is not provided (i.e. for local remotes), this index will
+            be kept in memory but not saved to disk.
```

Comment:
Should this go in a constructor docstring instead?

Reply:
afaik standard python conventions allow constructor args to be documented in either `__init__` or in the class declaration like here. I put them here because that's how it was done in some other dvc classes (see `dvc/cache.py::Cache`)